tests/scenarios/watch-mode-test.ts

```diff
@@ -15,20 +15,235 @@ let app = appScenarios.map('watch-mode', () => {
    */
 });
 
+abstract class Waiter {
+  readonly promise: Promise<void>;
+  protected _resolve!: () => void;
+  protected _reject!: (error: unknown) => void;
+  private _timeout = (timeout: number) => this.onTimeout(timeout);
+
+  constructor(timeout: number | null = 60000) {
+    this.promise = new Promise<void>((resolve, reject) => {
+      this._resolve = resolve;
+      this._reject = reject;
+    });
+
+    if (timeout !== null) {
+      setTimeout(() => this._timeout(timeout), timeout);
+    }
+  }
+
+  abstract onOutputLine(data: string): boolean;
+  abstract onExit(code: number): void;
+  abstract onTimeout(timeout: number): void;
+
+  protected resolve(): void {
+    const resolve = this._resolve;
+    this._resolve = this._reject = this._timeout = () => {};
+    resolve();
+  }
+
+  protected reject(error: unknown): void {
+    const reject = this._reject;
+    this._resolve = this._reject = this._timeout = () => {};
+    reject(error);
+  }
+}
+
+class OutputWaiter extends Waiter {
+  constructor(private process: EmberCLI, private output: string | RegExp) {
+    super();
+  }
+
+  onOutputLine(line: string): boolean {
+    if (this.matchLine(line)) {
+      this.resolve();
+      return true;
+    } else {
+      return false;
+    }
+  }
+
+  onExit(code: number): void {
+    try {
+      throw new Error(
+        'Process exited with code ' +
+          code +
+          ' before output "' +
+          this.output +
+          '" was found. ' +
+          'Recent output:\n\n' +
+          this.process.recentOutput
+      );
+    } catch (error) {
+      this.reject(error);
+    }
+  }
+
+  onTimeout(timeout: number): void {
+    try {
+      throw new Error(
+        'Timed out after ' +
+          timeout +
+          'ms before output "' +
+          this.output +
+          '" was found. ' +
+          'Recent output:\n\n' +
+          this.process.recentOutput
+      );
+    } catch (error) {
+      this.reject(error);
+    }
+  }
+
+  private matchLine(line: string): boolean {
+    if (typeof this.output === 'string') {
+      return this.output === line;
+    } else {
+      return this.output.test(line);
+    }
+  }
+}
+
+type Status = { type: 'starting' } | { type: 'ready' } | { type: 'errored'; error: unknown } | { type: 'completed' };
```

Comment:
It kinda feels like this whole util belongs in ember-cli and should have its own tests.
But that would mean ember-cli would need to formally support a js api, which could probably be implemented without stream peaking
:upside_down_face: idk.

Looks solid tho!

Reply:
I called it EmberCLI but it really is just a stateful child process thing with facilities for testing outputs, which I am sure could be useful for testing the vite integrations and what not as well